man/man1/pmiostat.1
'\"! tbl | mmdoc
'\"macro stdmacro
.\"
.\" Copyright (c) 2014 Red Hat Inc.  All Rights Reserved.
.\" 
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\" 
.\"
.TH PMIOSTAT 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmiostat\f1 \- performance metrics i/o statistics tool
.\" literals use .B or \f3
.\" arguments use .I or \f2
.SH SYNOPSIS
\f3pmiostat\f1
[\f3\-A\f1 \f2align\f1 \f3--align=\f2TIME\f1]
[\f3\-a\f1 \f2archive\f1 \f3--archive=\f2FILE\f1]
[\f3\-h\f1 \f2host\f1 \f3--host=\f2HOST\f1]
[\f3\-O\f1 \f2offset\f1 \f3--origin=\f2TIME\f1]
[\f3\-S\f1 \f2starttime\f1 \f3--start=\f2TIME\f1]
[\f3\-s\f1 \f2samples\f1 \f3--samples=\f2N\f1]
[\f3\-T\f1 \f2endtime\f1 \f3--finish=\f2TIME\f1]
[\f3\-t\f1 \f2interval\f1 \f3--interval=\f2DELTA\f1]
[\f3\-Z\f1 \f2timezone\f1 \f3--timezone=\f2TZ\f1]
[\f3\-z\f1 \f3--hostzone\f1]
[\f3\-?\f1 \f3--help\f1]
[\f3\-x\f1 [dm][,t][,h]\f1]
.SH DESCRIPTION
.de EX
.in +0.5i
.ie t .ft CB
.el .ft B
.ie t .sp .5v
.el .sp
.ta \\w' 'u*8
.nf
..
.de EE
.fi
.ie t .sp .5v
.el .sp
.ft R
.in
..
.B pmiostat
reports I/O statistics for scsi devices (by default) or device-mapper devices (if the \f3-x dm\f1 option is specified).
By default
.B pmiostat
reports live data for the local host but can also report for a remote host (\f3-h\fP) or from a previously captured PCP archive (\f3-a\fP).
.PP
The
.BR \-S ,
.BR \-T ,
.BR \-O
and
.B \-A
options may be used to define a time window to restrict the
samples retrieved, set an initial origin within the time window,
or specify a ``natural'' alignment of the sample times; refer to
.BR PCPIntro (1)
for a complete description of these options.
.PP
The other options which control the source, timing and layout of the information
reported by
.B pmiostat
are as follows:
.TP 5
.B \-a
Performance metric values are retrieved from the Performance Co-Pilot (PCP)
archive log file identified by the base name
.IR archive .
.TP
.B \-h
Current performance metric values are retrieved from the nominated
.I host
machine.
.TP
.B \-s
The argument
.I samples
defines the number of samples to be retrieved and reported.
If
.I samples
is 0 or
.B \-s
is not specified, 
.B pmiostat
will sample and report continuously (in real time mode) or until the end
of the PCP archive (in archive mode).
.TP
.B \-t
The default update \f2interval\f1 may be set to something other than the
default 1 second.
The
.I interval
argument follows the syntax described in
.BR PCPIntro (1),
and in the simplest form may be an unsigned integer (the implied
units in this case are seconds).
The \f3-t\fP option is particularly useful when replaying large archives (\f3-a\fP option) that span several hours or even days.
In this case specifying a large
.I interval
(e.g. 1h for 1 hour)
will reduce the volume of data reported and the i/o statistics will be averaged (interpolated) over
the reporting interval.
.TP
.B \-Z
By default,
.B pmiostat
reports the time of day according to the local timezone on the
system where
.B pmiostat
is run.
The
.B \-Z
option changes the timezone to
.I timezone
in the format of the environment variable
.B TZ
as described in
.BR environ (5).
.TP
.B \-z
Change the reporting timezone to the local timezone at the host that is
the source of the performance metrics, as identified via either the
.B \-h
or
.B \-a
options.
.TP
.B \-x
Specifies a comma separated list of one or more extended reporting options as follows:
.br
\f3dm\fP - report statistics for device-mapper logical devices instead of scsi devices,
.br
\f3t\fP - prefix every line in the report with a timestamp in \f2ctime\fP(3) format,
.br
\f3H\fP - omit the heading, which is otherwise reported every 24 samples.
.SH REPORT
The columns in the
.B pmiostat
report have the following interpretation :
.TP
.B Timestamp
When the \f3-x t\fP option is specified, this column is the timestamp in \f3ctime\fP(3) format.
.TP
.B Device
Specifies the scsi device name, or if \f3-x dm\fP is specified, the device-mapper logical device name.
.TP
.B rrqm/s
The number of read requests expressed as a rate per-second that were merged
during the reporting interval by the I/O scheduler.
.TP
.B wrqm/s
The number of write requests expressed as a rate per-second that were merged
during the reporting interval by the I/O scheduler.
.TP
.B r/s
The number of read requests completed by the device (after merges), expressed as a rate per second during the reporting interval.
.TP
.B w/s
The number of write requests completed by the device (after merges), expressed as a rate per second during the reporting interval.
.TP
.B rkB/s
The average volume of data read from the device expressed as KBytes/second during the reporting interval.
.TP
.B wkB/s
The average volume of data written to the device expressed as KBytes/second during the reporting interval.
.TP
.B avgrq-sz
The average I/O request size for both reads and writes to the device expressed as Kbytes during the reporting interval.
.TP
.B avgqu-sz
The average queue length of read and write requests to the device during the reporting interval.
.TP
.B await
The average time in milliseconds that read and write requests were queued (and serviced) to the device during the reporting interval.
.TP
.B r_await
The average time in milliseconds that read requests were queued (and serviced) to the device during the reporting interval.
.TP
.B w_await
The average time in milliseconds that write requests were queued (and serviced) to the device during the reporting interval.
.TP
.B %util
The percentage of time during the reporting interval that the device was busy processing requests.
A value of 100% indicates device saturation.
.SH FILES
.PD 0
.TP 10
.BI $PCP_VAR_DIR/pmns/ *
default PMNS specification files
.PD
.SH "PCP ENVIRONMENT"
Environment variables with the prefix
.B PCP_
are used to parameterize the file and directory names
used by PCP.
On each installation, the file
.I /etc/pcp.conf
contains the local values for these variables.
The
.B $PCP_CONF
variable may be used to specify an alternative
configuration file,
as described in
.BR pcp.conf (5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmcd (1),
.BR pmchart (1),
.BR pmdumplog (1),
.BR pminfo (1),
.BR pmlogger (1),
.BR pmtime (1),
.BR PMAPI (3),
.BR __pmSpecLocalPMDA (3),
.BR pcp.conf (5)
and
.BR pcp.env (5).
.SH DIAGNOSTICS
All are generated on standard error and are intended to be self-explanatory.


CHANGELOG
pcp-3.10.4 (15 April 2015)
    - pmdalinux: fixed kernel.pernode.cpu.* for systems with only one CPU.
    - pmdadm: merged in pmdadmcache, added thin provisioning metrics
    - pmcd: fix failure to recover from errors sending client attrs
    - pmcd: don't advertise presence if -A specified.
    - pmchart: fix metric/host selection for containers
    - python build: support for epel5 builds using python26 packages
    - python api: do not assume unicode is only for python3
    - stdpmid: updated the reserved domain number ranges
    - libpcp: fix instance request memory corruption for large requests
    - libpcp: multi-threading fixes and pdubuf optimisations
    - libpcp: fix Avahi-related memory corruption
    - libpcp: fix an interpolation performance issue
    - libpcp: fix getopt --hostsfile parsing
    - libpcp: tighten error handling in __pmGetArchiveEnd()
    - pmwebd: enable multi-threading by default
    - pmwebd: correct _indom array-allocation logic
    - pmwebd: additional ACAO headers to resource responses
    - pcp2graphite: python version neutral handling of stdout flush
    - pcp-verify: a new utility for verifying collector setups, especially
      for use with containers
    - pcp sh: fix tempfile leak with sub-commands, refactor pcp-summary
    - rc scripts: ensure rundir directories created when needed
    - rc scripts: ensure tmpfiles not leftover from helper scripts
    - rc_pmcd: add .NeedRemove option for pmcd start script also
    - pmpause: lightweight pmsleep extension for a pmcd container
    - pmlogcheck: merge previous versions, added documentation and tests
    - pmdumplog: new options and new log reading logic
    - build: add missing configure check for pkg-config requirement
    - build: Dockerfile for a pcp-pmcd container image based on Fedora
    - build: get Mac builds going on current Mac OS X (10.9/10.10)
    - configure: fix AC_MSG_ERROR parameter passing
    - packaging: fix default pmlogger/pmie configs directory location
    - packaging: simplify rpm dependencies on pcp-webjs package via %dir
    - packaging: relax the allowed-libpfm-version checks a little

pcp-3.10.3 (2 March 2015)
    - enhancement: ongoing improvements to the native containers support
	http://pcp.io/docs/lab.containers.html
    - enhancement: pmlogger on-the-fly one-shot recording
	http://oss.sgi.com/bugzilla/show_bug.cgi?id=1073
    - doc: containers tutorial.
    - build: new scripts directory. Contains useful scripts for project
      management tasks.
    - build: python-pcp no longer required.
    - build: python3 now used where available.
	https://bugzilla.redhat.com/show_bug.cgi?id=1194324
    - build: new --with-papi configure option.
    - qa: lots of improvements for robustness, portablity and determinism.
    - qa: duplicate PMNS names allowed by default.
    - pmdapapi: add papi refresh/read batching: papi.control.batch.
    - pmdalinux: improve /proc/cpuinfo parser for irregular formats.
    - services: ensure that they start after pmcd.
    - libpcp: allow duplicate names to map to the same PMID by default.
    - libpcp: duplicate PMNS names allowed by default.
    - pminfo: duplicate PMNS names allowed by default. Reverse the semantics of
      the -n and -N options.
    - pmlogger: duplicate PMNS names allowed by default
    - dbpmda: duplicate PMNS names allowed by default
    - pmcd: duplicate PMNS names allowed by default. Reverse the semantics of
      the -n and -N options.
    - pmdumplog: duplicate PMNS names allowed by default.
    - pmlogreduce: duplicate PMNS names allowed by default.
    - pmlogsummary: duplicate PMNS names allowed by default.
    - pmlogcheck: duplicate PMNS names allowed by default.
    - pmevent: duplicate PMNS names allowed by default.
    - pmchart: add an extra optional chart colour scheme.
    - solaris PMDA: duplicate PMNS names allowed by default.
    - ganglia2pcp: new utility.
    - bug fix: pmlogger: signal delivery may lead to deadlock
	http://oss.sgi.com/bugzilla/show_bug.cgi?id=1104
    - bug fix: libpcp: AF functionality has posix-signal-unsafe elements
	http://oss.sgi.com/bugzilla/show_bug.cgi?id=1069

pcp-3.10.2 (23 January 2015)
    - PMAPI extension: pmParseUnitsStr. Includes python binding
    - libpcp/pmcd: protocol-level support for containers
    - libpcp_pmda - add pmdaInterfaceMoved() function
    - pmdalinux: improve help text for network metrics
    - pmdalinux: remove support for devfs device names - replaced by udev
    - pmdalinux: rework hinv.map.scsi for longer scsi device names and indom
    - pmdalinux: switch to running as a daemon, as root
      persistence
    - activeMQ: new pmda
    - CIFS: new pmda (Common Internet File System)
    - pmda unbound: a new PMDA for the Unbound DNS resolver
    - pmdapapi: add configure conditional to test for
      PAPI_component_info_t->disabled
    - pmdapapi: add papi.available.version to output current PAPI version being
      used
    - pmdapapi: other new metrics
    - pmdapapi: drop root requirement for less sensitive papi.available metrics
    - pmdapapi: use dynamic metric description
    - dbpmda: extend string syntax to include both " and ' delimiters
    - rc_pmcd: execute _pmda_setup in the background
    - Solaris build improvements:
      - Support for Solaris symbol versioning
      - use a libcpp exports comment style that works for both Linux and Solaris
    - pmwebd: buildability fixes for illumos-based distros
    - pmda root: a privileged pmcd co-process
      Initial implementation of "pmdaroot", a helper co-process
      for pmcd that runs with elevated privileges and is able to
      (well, will be able to soon) offer services to pmcd and to
      other PMDAs where root access is required
    - pmdaroot/libpcp_pmda: interfaces for changing namespaces
    - pmlogextract robustness improvements
    - add pcp2graphite
    - fix security PR1101: unprivileged access to linux proc.* metrics on modern
      kernel

src/pmiostat/pmiostat.py
#!/usr/bin/python
#
# Copyright (C) 2014-2015 Red Hat.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Iostat Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# pylint: disable=C0103,R0914,R0902
""" Display disk and device-mapper I/O statistics """

import sys
from pcp import pmapi, pmcc
from cpmapi import PM_TYPE_U64, PM_CONTEXT_ARCHIVE, PM_SPACE_KBYTE

IOSTAT_SD_METRICS = [ 'disk.dev.read', 'disk.dev.read_bytes',
                 'disk.dev.write', 'disk.dev.write_bytes',
                 'disk.dev.read_merge', 'disk.dev.write_merge',
                 'disk.dev.blkread', 'disk.dev.blkwrite',
                 'disk.dev.read_rawactive', 'disk.dev.write_rawactive',
                 'disk.dev.avactive']

IOSTAT_DM_METRICS = [ 'disk.dm.read', 'disk.dm.read_bytes',
                 'disk.dm.write', 'disk.dm.write_bytes',
                 'disk.dm.read_merge', 'disk.dm.write_merge',
                 'disk.dm.blkread', 'disk.dm.blkwrite',
                 'disk.dm.read_rawactive', 'disk.dm.write_rawactive',
                 'disk.dm.avactive']

class IostatReport(pmcc.MetricGroupPrinter):
    Hcount = 0
    def timeStampDelta(self, group):
        c = 1000000.0 * group.timestamp.tv_sec + group.timestamp.tv_usec
        p = 1000000.0 * group.prevTimestamp.tv_sec + group.prevTimestamp.tv_usec
        return (c - p) / 1000000.0

    def instlist(self, group, name):
        return dict(map(lambda x: (x[1], x[2]), group[name].netValues)).keys()

    def curVals(self, group, name):
        return dict(map(lambda x: (x[1], x[2]), group[name].netValues))

    def prevVals(self, group, name):
        return dict(map(lambda x: (x[1], x[2]), group[name].netPrevValues))

    def report(self, manager):
        if 'dm' in IostatOptions.xflag:
            subtree = 'disk.dm'
        else:
            subtree = 'disk.dev'
        group = manager["iostat"]

        if group[subtree + '.read_merge'].netPrevValues == None:
            # need two fetches to report rate converted counter metrics
            return

        instlist = self.instlist(group, subtree + '.read')
        dt = self.timeStampDelta(group)
        timestamp = group.contextCache.pmCtime(int(group.timestamp)).rstrip()

        c_rrqm = self.curVals(group, subtree + '.read_merge')
        p_rrqm = self.prevVals(group, subtree + '.read_merge')

        c_wrqm = self.curVals(group, subtree + '.write_merge')
        p_wrqm = self.prevVals(group, subtree + '.write_merge')

        c_r = self.curVals(group, subtree + '.read')
        p_r = self.prevVals(group, subtree + '.read')

        c_w = self.curVals(group, subtree + '.write')
        p_w = self.prevVals(group, subtree + '.write')

        c_rkb = self.curVals(group, subtree + '.read_bytes')
        p_rkb = self.prevVals(group, subtree + '.read_bytes')

        c_wkb = self.curVals(group, subtree + '.write_bytes')
        p_wkb = self.prevVals(group, subtree + '.write_bytes')

        c_ractive = self.curVals(group, subtree + '.read_rawactive')
        p_ractive = self.prevVals(group, subtree + '.read_rawactive')

        c_wactive = self.curVals(group, subtree + '.write_rawactive')
        p_wactive = self.prevVals(group, subtree + '.write_rawactive')

        c_avactive = self.curVals(group, subtree + '.avactive')
        p_avactive = self.prevVals(group, subtree + '.avactive')

        # check availability
        if p_rrqm == {} or p_wrqm == {} or p_r == {} or p_w == {} or p_rkb == {} \
           or p_wkb == {} or p_ractive == {} or p_wactive == {} or p_avactive == {}:
            # no values (near start of archive?)
            return

        if "h" not in IostatOptions.xflag:
            self.Hcount += 1
            if self.Hcount == 24:
                self.Hcount = 1
            if self.Hcount == 1:
                if "t" in IostatOptions.xflag:
                    heading = ('# Timestamp', 'Device', 'rrqm/s', 'wrqm/s', 'r/s', 'w/s', 'rkB/s', 'wkB/s',
                               'avgrq-sz', 'avgqu-sz', 'await', 'r_await', 'w_await', '%util')
                    print("%-24s %-12s %7s %7s %6s %6s %8s %8s %8s %8s %7s %7s %7s %5s" % heading)
                else:
                    heading = ('# Device', 'rrqm/s', 'wrqm/s', 'r/s', 'w/s', 'rkB/s', 'wkB/s',
                               'avgrq-sz', 'avgqu-sz', 'await', 'r_await', 'w_await', '%util')
                    print("%-12s %7s %7s %6s %6s %8s %8s %8s %8s %7s %7s %7s %5s" % heading)

        for inst in sorted(instlist):
            # basic stats
            rrqm = (c_rrqm[inst] - p_rrqm[inst]) / dt
            wrqm = (c_wrqm[inst] - p_wrqm[inst]) / dt
            r = (c_r[inst] - p_r[inst]) / dt
            w = (c_w[inst] - p_w[inst]) / dt
            rkb = (c_rkb[inst] - p_rkb[inst]) / dt
            wkb = (c_wkb[inst] - p_wkb[inst]) / dt

            # totals
            tot_rios = (float)(c_r[inst] - p_r[inst])
            tot_wios = (float)(c_w[inst] - p_w[inst])
            tot_ios = (float)(tot_rios + tot_wios)

            # total active time in seconds (same units as dt)
            tot_active = (float)(c_avactive[inst] - p_avactive[inst]) / 1000.0

            avgrqsz = avgqsz = await = r_await = w_await = util = 0.0

            # average request size units are KB (sysstat reports in units of sectors)
            if tot_ios:
                avgrqsz = (float)((c_rkb[inst] - p_rkb[inst]) + (c_wkb[inst] - p_wkb[inst])) / tot_ios

            # average queue length
            avgqsz = (float)((c_ractive[inst] - p_ractive[inst]) + (c_wactive[inst] - p_wactive[inst])) / dt / 1000.0

            # await, r_await, w_await
            if tot_ios:
                await = ((c_ractive[inst] - p_ractive[inst]) + (c_wactive[inst] - p_wactive[inst])) / tot_ios

            if tot_rios:
                r_await = (c_ractive[inst] - p_ractive[inst]) / tot_rios

            if tot_wios:
                w_await = (c_wactive[inst] - p_wactive[inst]) / tot_wios

            # device utilization (percentage of active time / interval)
            if tot_active:
                    util = 100.0 * tot_active / dt

            device = inst	# prepare name for printing
            if "t" in IostatOptions.xflag:
                print("%-24s %-12s %7.1f %7.1f %6.1f %6.1f %8.1f %8.1f %8.2f %8.2f %7.1f %7.1f %7.1f %5.1f" \
                % (timestamp, device, rrqm, wrqm, r, w, rkb, wkb, avgrqsz, avgqsz, await, r_await, w_await, util))
            else:
                print("%-12s %7.1f %7.1f %6.1f %6.1f %8.1f %8.1f %8.2f %8.2f %7.1f %7.1f %7.1f %5.1f" \
                % (device, rrqm, wrqm, r, w, rkb, wkb, avgrqsz, avgqsz, await, r_await, w_await, util))


class IostatOptions(pmapi.pmOptions):
    # class attributes
    xflag = [] 

    def extraOptions(self, opt, optarg, index):
        if opt == "x":
            IostatOptions.xflag = optarg.replace(',', ' ').split(' ')
        else:
            print("Warning: option '%s' not recognised" % opt)

    def __init__(self):
        pmapi.pmOptions.__init__(self, "A:a:D:h:O:S:s:T:t:VZ:z?x:")
        self.pmSetOptionCallback(self.extraOptions)
        self.pmSetLongOptionHeader("General options")
        self.pmSetLongOptionAlign()
        self.pmSetLongOptionArchive()
        self.pmSetLongOptionDebug()
        self.pmSetLongOptionHost()
        self.pmSetLongOptionOrigin()
        self.pmSetLongOptionStart()
        self.pmSetLongOptionSamples()
        self.pmSetLongOptionFinish()
        self.pmSetLongOptionInterval()
        self.pmSetLongOptionVersion()
        self.pmSetLongOptionTimeZone()
        self.pmSetLongOptionHostZone()
        self.pmSetLongOptionHelp()
        self.pmSetLongOptionHeader("Extended options")
        self.pmSetLongOption("", 1, 'x', "LIST", "comma separated extended options: [[dm],[t],[h]]")
        self.pmSetLongOptionText("\t\tdm\tshow device-mapper statistics (default is sd devices)")
        self.pmSetLongOptionText("\t\tt\tprecede every line with a timestamp in ctime format");
        self.pmSetLongOptionText("\t\th\tsuppress headings");

if __name__ == '__main__':
    try:
        manager = pmcc.MetricGroupManager.builder(IostatOptions(), sys.argv)
        if "dm" in IostatOptions.xflag :
            manager["iostat"] = IOSTAT_DM_METRICS
        else:
            manager["iostat"] = IOSTAT_SD_METRICS
        manager.printer = IostatReport()
        sts = manager.run()
        sys.exit(sts)
    except pmapi.pmErr as error:
        print('%s: %s\n' % (error.progname(), error.message()))
    except pmapi.pmUsageErr as usage:
        usage.message()
        sys.exit(1)
    except KeyboardInterrupt:
        pass
